### PR TITLE
:sparkles: add spotify-api (don't approve)

### PR DESCRIPTION
**Pull request by Mimaaa**
Signed-off-by: Mirza van Meerwijk <mirzavanmeerwijk@gmail.com>

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "Code &amp; documentation for the Minor Class: Web Of Things 📡",
   "main": "app.js",
   "scripts": {
-    "start": "npm i && node app.js",
+    "start": "node app.js",
```
Comment (dandevri) [inline package.json]: Please revert back to old start script 🎉

**Review by dandevri — CHANGES_REQUESTED** [review]

**Review by dandevri — APPROVED** [review]
🌮

File: views/current.ejs [hunk 2]
```diff
@@ -25,7 +25,6 @@
         <button formaction="/play">▶</button>
```
Comment (IanCStewart) [inline views/current.ejs]: These buttons can go

**Review by IanCStewart — CHANGES_REQUESTED** [review]

**Review by IanCStewart — CHANGES_REQUESTED** [review]
Add `skip` function to SpotifyEndpoint

File: app.js [hunk 3]
```diff
@@ -1,25 +1,174 @@
...
+  // Set HTTP header with the STATE_KEY and generated string
+  res.cookie(STATE_KEY, state);
+  // Create authorization URL with the scopes and state in order to receive correct data
+  res.redirect(spotifyApi.createAuthorizeURL(scopes, state));
+}
 
-app.get('/api/vote', function (req, res) {
+function callback (req, res) {
+  const { code, state } = req.query;
+  var storedState = req.cookies ? req.cookies[STATE_KEY] : null;
```
Comment (IanCStewart) [inline app.js]: try this:
```
var storedSate = null;

if (req.cookies) {
  storedSate = req.cookies[STATE_KEY];
}
```

File: app.js [hunk 4]
```diff
@@ -1,25 +1,174 @@
...
+     res.redirect('/#' +
+       querystring.stringify({
+         error: 'state_mismatch'
+       }));
+  // If the state is valid, get the authorization code and pass it on the client
+  } else {
+     // Clear the cookie that was set
+     res.clearCookie(STATE_KEY);
+     // Retrieve an access token and a refresh token
+     spotifyApi.authorizationCodeGrant(code).then(data => {
```
Comment (IanCStewart) [inline app.js]: Pretty sure we agreed to use es5 instead of es6. So no fat arrows for us 😢

File: package.json [hunk 5]
```diff
@@ -20,10 +20,15 @@
...
   "dependencies": {
+    "cookie-parser": "^1.4.3",
+    "dotenv": "^4.0.0",
     "ejs": "^2.5.6",
     "express": "^4.15.2",
     "lodash": "^4.17.4",
-    "node-fetch": "^1.6.3"
+    "node-fetch": "^1.6.3",
+    "querystring": "^0.2.0",
+    "request": "^2.81.0",
```
Comment (IanCStewart) [inline package.json:30]: looks like we are using two different `npm` modules to do an HTTP request. We should probably fix this.

File: views/current.ejs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      <hr>
+      <div class="voting">
+        <dt>Users voted:</dt>
+        <dd>👍 1</dd>
+      </div>
+    </dl>
+
+    <div class="controls">
+      <form method="post">
+        <button formaction="/play">▶</button>
```
Comment (IanCStewart) [inline views/current.ejs:24]: Maybe nice if we use an emoji here?

**Review by IanCStewart — CHANGES_REQUESTED** [review]

File: views/current.ejs (hunk 6 above)
Comment (dandevri) [reply views/current.ejs:24]: Or UTF-8 characters?

File: app.js (hunk 3 above)
Comment (Mimaaa) [reply app.js]: doesn't work

File: views/current.ejs (hunk 6 above)
Comment (Mimaaa) [reply views/current.ejs:24]: fixed

**Review by dandevri — COMMENTED** [review]
🌮

**Pull request merged**